processing/src/main/java/io/druid/query/lookup/LookupReferencesManager.java

```diff
@@ -52,333 +56,313 @@
+  @VisibleForTesting
+  interface Notice
   {
-    try {
-      startStopLock.readLock().lockInterruptibly();
-    }
-    catch (InterruptedException e) {
-      throw Throwables.propagate(e);
-    }
-    try {
-      final LookupExtractorFactory lookupExtractorFactory = lookupMap.get(lookupName);
-      assertStarted();
-      return lookupExtractorFactory;
-    }
-    finally {
-      startStopLock.readLock().unlock();
-    }
+    void handle();
   }
 
-  /**
-   * @return Returns {@link Map} containing a copy of the current state.
-   *
-   * @throws ISE if the is is closed or did not start yet.
-   */
-  public Map<String, LookupExtractorFactory> getAll()
+  private class LoadNotice implements Notice
   {
-    try {
-      startStopLock.readLock().lockInterruptibly();
-    }
-    catch (InterruptedException e) {
-      throw Throwables.propagate(e);
-    }
-    try {
-      assertStarted();
-      return Maps.newHashMap(lookupMap);
+    String lookupName;
+    LookupExtractorFactoryContainer lookupExtractorFactoryContainer;
+
+    public LoadNotice(String lookupName, LookupExtractorFactoryContainer lookupExtractorFactoryContainer)
+    {
+      this.lookupName = lookupName;
+      this.lookupExtractorFactoryContainer = lookupExtractorFactoryContainer;
     }
-    finally {
-      startStopLock.readLock().unlock();
+
+    @Override
+    public void handle()
+    {
+      Preconditions.checkState(lifecycleLock.isStarted());
+
+      LookupExtractorFactoryContainer old = lookupMap.get(lookupName);
+      if (old != null && !lookupExtractorFactoryContainer.replaces(old)) {
+        LOG.warn(
+            "got notice to load lookup [%s] that can't replace existing [%s].",
+            lookupExtractorFactoryContainer,
+            old
+        );
+        return;
+      }
+
+      if (!lookupExtractorFactoryContainer.getLookupExtractorFactory().start()) {
+        throw new ISE("start method returned false for lookup [%s]:[%s]", lookupName, lookupExtractorFactoryContainer);
+      }
+
+      old = lookupMap.put(lookupName, lookupExtractorFactoryContainer);
+
+      if (LOG.isDebugEnabled()) {
+        LOG.debug("Loaded lookup [%s] with spec [%s].", lookupName, lookupExtractorFactoryContainer);
+      }
+
+      takeSnapshot();
+
+      if (old != null) {
+        if (!old.getLookupExtractorFactory().close()) {
+          throw new ISE("close method returned false for lookup [%s]:[%s]", lookupName, old);
+        }
+      }
     }
   }
 
-  private void assertStarted() throws ISE
+  private class DropNotice implements Notice
   {
-    if (isClosed()) {
-      throw new ISE("lookup manager is closed");
+    String lookupName;
+
+    public DropNotice(String lookupName)
+    {
+      this.lookupName = lookupName;
     }
-  }
 
-  public boolean isClosed()
-  {
-    return !started.get();
-  }
+    @Override
+    public void handle()
+    {
+      Preconditions.checkState(lifecycleLock.isStarted());
```

Comment:
This check is not needed because one is already done in the main loop in mainThread.